--- before/net/minecraft/client/gui/GuiScreenBook.java
+++ after/net/minecraft/client/gui/GuiScreenBook.java
@@ -1,5 +1,6 @@
 package net.minecraft.client.gui;
 
+import com.cleanroommc.client.ime.IMEWrapperBook;
 import com.google.common.collect.Lists;
 import com.google.gson.JsonParseException;
 import io.netty.buffer.Unpooled;
@@ -27,12 +28,11 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.lwjgl.input.Keyboard;
+import org.lwjglx.input.Keyboard;
 
 @SideOnly(Side.CLIENT)
 public class GuiScreenBook extends GuiScreen
 {
-    private static final Logger field_146473_a = LogManager.getLogger();
     private static final ResourceLocation field_146466_f = new ResourceLocation("textures/gui/book.png");
     private final EntityPlayer field_146468_g;
     private final ItemStack field_146474_h;
@@ -69,6 +69,7 @@
 
             if (this.field_146476_w < 1)
             {
+                this.field_146483_y.func_74742_a(new NBTTagString("")); // Forge: fix MC-1685
                 this.field_146476_w = 1;
             }
         }
@@ -98,6 +99,9 @@
             this.field_146472_C = this.func_189646_b(new GuiButton(0, this.field_146294_l / 2 + 2, 196, 98, 20, I18n.func_135052_a("gui.done")));
             this.field_146467_E = this.func_189646_b(new GuiButton(5, this.field_146294_l / 2 - 100, 196, 98, 20, I18n.func_135052_a("book.finalizeButton")));
             this.field_146469_F = this.func_189646_b(new GuiButton(4, this.field_146294_l / 2 + 2, 196, 98, 20, I18n.func_135052_a("gui.cancel")));
+            IMEWrapperBook.setBookGui(this);
+            IMEWrapperBook.setText(func_146456_p());
+            IMEWrapperBook.instance.setVisible(true);
         }
         else
         {
@@ -105,15 +109,15 @@
         }
 
         int i = (this.field_146294_l - 192) / 2;
-        int j = 2;
-        this.field_146470_A = (GuiScreenBook.NextPageButton)this.func_189646_b(new GuiScreenBook.NextPageButton(1, i + 120, 156, true));
-        this.field_146471_B = (GuiScreenBook.NextPageButton)this.func_189646_b(new GuiScreenBook.NextPageButton(2, i + 38, 156, false));
+        this.field_146470_A = this.func_189646_b(new NextPageButton(1, i + 120, 156, true));
+        this.field_146471_B = this.func_189646_b(new NextPageButton(2, i + 38, 156, false));
         this.func_146464_h();
     }
 
     public void func_146281_b()
     {
         Keyboard.enableRepeatEvents(false);
+        IMEWrapperBook.instance.setVisible(false);
     }
 
     private void func_146464_h()
@@ -131,7 +135,7 @@
         }
     }
 
-    private void func_146462_a(boolean p_146462_1_) throws IOException
+    private void func_146462_a(boolean p_146462_1_)
     {
         if (this.field_146475_i && this.field_146481_r)
         {
@@ -214,7 +218,7 @@
             else if (p_146284_1_.field_146127_k == 5 && this.field_146480_s)
             {
                 this.func_146462_a(true);
-                this.field_146297_k.func_147108_a((GuiScreen)null);
+                this.field_146297_k.func_147108_a(null);
             }
             else if (p_146284_1_.field_146127_k == 4 && this.field_146480_s)
             {
@@ -260,62 +264,48 @@
         }
         else
         {
-            switch (p_146463_2_)
-            {
-                case 14:
+            switch (p_146463_2_) {
+                case 14 -> {
                     String s = this.func_146456_p();
-
-                    if (!s.isEmpty())
-                    {
+                    if (!s.isEmpty()) {
                         this.func_146457_a(s.substring(0, s.length() - 1));
                     }
-
-                    return;
-                case 28:
-                case 156:
+                }
+                case 28, 156 -> {
                     this.func_146459_b("\n");
-                    return;
-                default:
-
-                    if (ChatAllowedCharacters.func_71566_a(p_146463_1_))
-                    {
+                    IMEWrapperBook.addReturn();
+                }
+                default -> {
+                    if (ChatAllowedCharacters.func_71566_a(p_146463_1_)) {
                         this.func_146459_b(Character.toString(p_146463_1_));
                     }
+                }
             }
         }
     }
 
-    private void func_146460_c(char p_146460_1_, int p_146460_2_) throws IOException
+    private void func_146460_c(char p_146460_1_, int p_146460_2_)
     {
-        switch (p_146460_2_)
-        {
-            case 14:
-
-                if (!this.field_146482_z.isEmpty())
-                {
+        switch (p_146460_2_) {
+            case 14 -> {
+                if (!this.field_146482_z.isEmpty()) {
                     this.field_146482_z = this.field_146482_z.substring(0, this.field_146482_z.length() - 1);
                     this.func_146464_h();
                 }
-
-                return;
-            case 28:
-            case 156:
-
-                if (!this.field_146482_z.isEmpty())
-                {
+            }
+            case 28, 156 -> {
+                if (!this.field_146482_z.isEmpty()) {
                     this.func_146462_a(true);
-                    this.field_146297_k.func_147108_a((GuiScreen)null);
+                    this.field_146297_k.func_147108_a(null);
                 }
-
-                return;
-            default:
-
-                if (this.field_146482_z.length() < 16 && ChatAllowedCharacters.func_71566_a(p_146460_1_))
-                {
-                    this.field_146482_z = this.field_146482_z + Character.toString(p_146460_1_);
+            }
+            default -> {
+                if (this.field_146482_z.length() < 16 && ChatAllowedCharacters.func_71566_a(p_146460_1_)) {
+                    this.field_146482_z = this.field_146482_z + p_146460_1_;
                     this.func_146464_h();
                     this.field_146481_r = true;
                 }
+            }
         }
     }
 
@@ -345,12 +335,23 @@
         }
     }
 
+    public void pageSetCurrentNoSync(String s) {
+        int i = this.field_146289_q.func_78267_b(s + "" + TextFormatting.BLACK + "_", 118);
+
+        if (i <= 128 && s.length() < 256)
+        {
+            if (this.field_146483_y != null && this.field_146484_x >= 0 && this.field_146484_x < this.field_146483_y.func_74745_c())
+            {
+                this.field_146483_y.func_150304_a(this.field_146484_x, new NBTTagString(s));
+                this.field_146481_r = true;
+            }
+        }
+    }
     public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
     {
         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
         this.field_146297_k.func_110434_K().func_110577_a(field_146466_f);
         int i = (this.field_146294_l - 192) / 2;
-        int j = 2;
         this.func_73729_b(i, 2, 0, 0, 192, 192);
 
         if (this.field_146480_s)
@@ -492,6 +493,7 @@
                 {
                     this.field_146484_x = i;
                     this.func_146464_h();
+                    IMEWrapperBook.setText(func_146456_p());
                     return true;
                 }
             }
@@ -508,7 +510,7 @@
 
             if (flag && clickevent.func_150669_a() == ClickEvent.Action.RUN_COMMAND)
             {
-                this.field_146297_k.func_147108_a((GuiScreen)null);
+                this.field_146297_k.func_147108_a(null);
             }
 
             return flag;
